File: src/Core/StagingFinancialEvents/Services/Transformers/EventProcessors/CashTransactionEventProcessor.py
import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.FinancialEvents.Schemas.CommonFormats import (
    GenericMonetaryExchangeInformation,
)
from Core.FinancialEvents.Schemas.Events import (
    TradeEventCashTransactionDividend,
    TradeEventCashTransactionPaymentInLieuOfDividend,
    TradeEventCashTransactionWitholdingTax,
    TradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividend,
)
from Core.FinancialEvents.Schemas.FinancialIdentifier import FinancialIdentifier
from Core.StagingFinancialEvents.Contracts.EventProcessor import EventProcessor
from Core.StagingFinancialEvents.Schemas.Events import (
    StagingTradeEventCashTransaction,
    StagingTradeEventCashTransactionDividend,
    StagingTradeEventCashTransactionPaymentInLieuOfDividends,
    StagingTradeEventCashTransactionWitholdingTax,
    StagingTradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividends,
)


class CashTransactionEventProcessor(EventProcessor[StagingTradeEventCashTransaction, pgf.TransactionCash]):

    def process(self, input: StagingTradeEventCashTransaction) -> pgf.TransactionCash:
        converted: pgf.TransactionCash

        if isinstance(input, StagingTradeEventCashTransactionDividend):
            converted = TradeEventCashTransactionDividend(
                ID=input.ID,
                FinancialIdentifier=FinancialIdentifier.fromStagingIdentifier(input.FinancialIdentifier),
                AssetClass=input.AssetClass,
                Date=input.Date,
                Multiplier=input.Multiplier,
                ExchangedMoney=GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=input.ExchangedMoney.UnderlyingQuantity,
                    UnderlyingTradePrice=input.ExchangedMoney.UnderlyingTradePrice,
                    UnderlyingCurrency=input.ExchangedMoney.UnderlyingCurrency,
                    ComissionCurrency=input.ExchangedMoney.ComissionCurrency,
                    ComissionTotal=input.ExchangedMoney.ComissionTotal,
                    TaxCurrency=input.ExchangedMoney.TaxCurrency,
                    TaxTotal=input.ExchangedMoney.TaxTotal,
                    FxRateToBase=input.ExchangedMoney.FxRateToBase,
                ),
                ActionID=input.ActionID,
                TransactionID=input.TransactionID,
                ListingExchange=input.ListingExchange,
                DividendType=input.DividendType,
            )
            return converted

        if isinstance(input, StagingTradeEventCashTransactionWitholdingTax):
            converted = TradeEventCashTransactionWitholdingTax(
                ID=input.ID,
                FinancialIdentifier=FinancialIdentifier.fromStagingIdentifier(input.FinancialIdentifier),
                AssetClass=input.AssetClass,
                Date=input.Date,
                Multiplier=input.Multiplier,
                ExchangedMoney=GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=input.ExchangedMoney.UnderlyingQuantity,
                    UnderlyingTradePrice=input.ExchangedMoney.UnderlyingTradePrice,
                    UnderlyingCurrency=input.ExchangedMoney.UnderlyingCurrency,
                    ComissionCurrency=input.ExchangedMoney.ComissionCurrency,
                    ComissionTotal=input.ExchangedMoney.ComissionTotal,
                    TaxCurrency=input.ExchangedMoney.TaxCurrency,
                    TaxTotal=input.ExchangedMoney.TaxTotal,
                    FxRateToBase=input.ExchangedMoney.FxRateToBase,
                ),
                ActionID=input.ActionID,
                TransactionID=input.TransactionID,
                ListingExchange=input.ListingExchange,
            )
            return converted

        if isinstance(input, StagingTradeEventCashTransactionPaymentInLieuOfDividends):
            converted = TradeEventCashTransactionPaymentInLieuOfDividend(
                ID=input.ID,
                FinancialIdentifier=FinancialIdentifier.fromStagingIdentifier(input.FinancialIdentifier),
                AssetClass=input.AssetClass,
                Date=input.Date,
                Multiplier=input.Multiplier,
                ExchangedMoney=GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=input.ExchangedMoney.UnderlyingQuantity,
                    UnderlyingTradePrice=input.ExchangedMoney.UnderlyingTradePrice,
                    UnderlyingCurrency=input.ExchangedMoney.UnderlyingCurrency,
                    ComissionCurrency=input.ExchangedMoney.ComissionCurrency,
                    ComissionTotal=input.ExchangedMoney.ComissionTotal,
                    TaxCurrency=input.ExchangedMoney.TaxCurrency,
                    TaxTotal=input.ExchangedMoney.TaxTotal,
                    FxRateToBase=input.ExchangedMoney.FxRateToBase,
                ),
                ActionID=input.ActionID,
                TransactionID=input.TransactionID,
                ListingExchange=input.ListingExchange,
                DividendType=input.DividendType,
            )
            return converted

        if isinstance(input, StagingTradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividends):
            converted = TradeEventCashTransactionWitholdingTaxForPaymentInLieuOfDividend(
                ID=input.ID,
                FinancialIdentifier=FinancialIdentifier.fromStagingIdentifier(input.FinancialIdentifier),
                AssetClass=input.AssetClass,
                Date=input.Date,
                Multiplier=input.Multiplier,
                ExchangedMoney=GenericMonetaryExchangeInformation(
                    UnderlyingQuantity=input.ExchangedMoney.UnderlyingQuantity,
                    UnderlyingTradePrice=input.ExchangedMoney.UnderlyingTradePrice,
                    UnderlyingCurrency=input.ExchangedMoney.UnderlyingCurrency,
                    ComissionCurrency=input.ExchangedMoney.ComissionCurrency,
                    ComissionTotal=input.ExchangedMoney.ComissionTotal,
                    TaxCurrency=input.ExchangedMoney.TaxCurrency,
                    TaxTotal=input.ExchangedMoney.TaxTotal,
                    FxRateToBase=input.ExchangedMoney.FxRateToBase,
                ),
                ActionID=input.ActionID,
                TransactionID=input.TransactionID,
                ListingExchange=input.ListingExchange,
            )
            return converted

        raise ValueError("Unknown Cash Transaction Staging type")

File: lib/src/Core/StagingProcessor/Transformers/CashTransactionEventProcessor.ts
import {
	TradeEventCashTransactionDividend,
	TradeEventCashTransactionPaymentInLieuOfDividend,
	TradeEventCashTransactionWithholdingTax,
	TradeEventCashTransactionWithholdingTaxForPaymentInLieuOfDividend,
	type TransactionCash,
} from "@brrr/Core/Schemas/Events";
import { FinancialIdentifier } from "@brrr/Core/Schemas/FinancialIdentifier";
import {
	StagingTradeEventCashTransactionDividend,
	StagingTradeEventCashTransactionPaymentInLieuOfDividends,
	StagingTradeEventCashTransactionWithholdingTax,
	type StagingTransactionCash,
} from "@brrr/Core/Schemas/Staging/Events";

export function processCashTransaction(input: StagingTransactionCash): TransactionCash {
	const identifier = FinancialIdentifier.fromStagingIdentifier(input.financialIdentifier);
	const base = {
		id: input.id,
		financialIdentifier: identifier,
		assetClass: input.assetClass,
		date: input.date,
		multiplier: input.multiplier,
		exchangedMoney: { ...input.exchangedMoney },
		actionId: input.actionId,
		transactionId: input.transactionId,
		provenance: [] as never[],
	};

	if (input instanceof StagingTradeEventCashTransactionDividend) {
		return new TradeEventCashTransactionDividend({
			...base,
			listingExchange: input.listingExchange,
			dividendType: input.dividendType,
		});
	} else if (input instanceof StagingTradeEventCashTransactionWithholdingTax) {
		return new TradeEventCashTransactionWithholdingTax({
			...base,
			listingExchange: input.listingExchange,
		});
	} else if (input instanceof StagingTradeEventCashTransactionPaymentInLieuOfDividends) {
		return new TradeEventCashTransactionPaymentInLieuOfDividend({
			...base,
			listingExchange: input.listingExchange,
			dividendType: input.dividendType,
		});
	} else {
		return new TradeEventCashTransactionWithholdingTaxForPaymentInLieuOfDividend({
			...base,
			listingExchange: input.listingExchange,
		});
	}
}
